src/Propel/Generator/Schema/SchemaParserVisitor.php

```diff
@@ -0,0 +1,414 @@
+    
+    /** @var Schema */
+    protected $schema;
+    
+    /** @var BehaviorManager */
+    protected $behaviorManager;
+    
+    /** @var SchemaReader */
+    protected $reader;
+    
+    private $allowedTags = [
+        'database',
+        'entity',
+        'table',
+        'field',
+        'column',
+        'relation',
+        'foreign-key',
+        'index',
+        'unique',
+        'behavior',
+        'vendor'
+    ];
+    
+    public function __construct(SchemaReader $reader) {
+        $this->reader = $reader;
+        $this->behaviorManager = new BehaviorManager();
+        $this->tagStack = new Stack();
+    }
+    
+    public function setSchema(Schema $schema) {
+        $this->schema = $schema;
+        $this->behaviorManager->setGeneratorConfig($schema->getGeneratorConfig());
+    }
+
+    public function visitElementStart($name, $attributes, $line, $column) {
+        $parentTag = $this->tagStack->peek();
+
+        // root
+        if (null === $parentTag) {
+            switch ($name) {
+                case 'database':
+                    if ($this->schema->isExternalSchema()) {
+                        $this->currDatabase = $this->schema->getRootSchema()->getDatabase();
+                    } else {
+                        $db = ModelFactory::createMappingModel($name, $attributes);
+                        $this->schema->addDatabase($db);
+                        $this->currDatabase = $db;
+                    }
+                    break;
+                    
+                default:
+                    $this->throwInvalidTagException($name, $line, $column);
+            }
+        }
+        
+        // regonized parent tag
+        else if (in_array($parentTag, $this->allowedTags)) {
+            $methodName = 'visit' . NamingTool::toUpperCamelCase($parentTag);
+            if (method_exists($this, $methodName)) {
+                $this->$methodName($name, $attributes, $line, $column);
+            }
+        }
+        
+        // at this point, it must be an invalid tag
+        else {
+            $this->throwInvalidTagException($name, $line, $column);
+        }
+        
+        $this->tagStack->push($name);
+    }
+    
+    protected function visitDatabase($name, $attributes, $line, $column) {
+        switch ($name) {
+            case 'external-schema':
+                $schema = new Schema();
+                
+                if (!$this->schema->isExternalSchema()) {
+                    $isRefOnly = $attributes['referenceOnly'] ?? null;
+                    $isRefOnly = null !== $isRefOnly ? ('true' === strtolower($isRefOnly)) : true;
+                    $schema->setReferenceOnly($isRefOnly);
+                }
+                
+                $filename = $attributes['filename'] ?? null;
+                
+                // very effective absolute path checking :D
+                // if not absolute, make it relative to the current schema
+                if ('/' !== $filename{0}) {
```

Comment:
the `very effective absolute path checking` ;)  is not yet capable of that. Windows should be supported, for sure. I was hoping for an `is_absolute()` function by php, which isn't. It's on my list.